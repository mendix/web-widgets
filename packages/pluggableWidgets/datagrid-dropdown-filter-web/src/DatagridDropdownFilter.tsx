import { useFilterContextValue } from "@mendix/pluggable-widgets-commons/components/web";
import { createElement, Fragment, ReactElement } from "react";
import { DatagridDropdownFilterContainerProps } from "../typings/DatagridDropdownFilterProps";
import { EnumerationFilter } from "./components/EnumerationFilter";
import { ErrorBox } from "./components/ErrorBox";
import { AssociationFilter } from "./components/AssociationFilter";
import { OutOfProviderError } from "./features/errors";

export default function DatagridDropdownFilter(props: DatagridDropdownFilterContainerProps): ReactElement {
    const context = useFilterContextValue();

    if (context.hasError) {
        return <ErrorBox error={new OutOfProviderError()} />;
    }

    if (props.defaultValue?.status === "loading") {
        return <Fragment />;
    }

    const Filter = context.value.associationProperties ? AssociationFilter : EnumerationFilter;

    return <Filter context={context.value} widgetProps={props} />;
}
