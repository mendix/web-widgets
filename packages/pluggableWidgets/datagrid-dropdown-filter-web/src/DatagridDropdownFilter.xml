<?xml version="1.0" encoding="utf-8" ?>
<widget id="com.mendix.widget.web.datagriddropdownfilter.DatagridDropdownFilter" pluginWidget="true" offlineCapable="true" xmlns="http://www.mendix.com/widget/1.0/" xmlns:xsi="http://www.w3.org/2001/XMLSchema-instance" xsi:schemaLocation="http://www.mendix.com/widget/1.0/ ../node_modules/mendix/custom_widget.xsd">
    <name>Drop-down filter</name>
    <description />
    <studioProCategory>Data controls</studioProCategory>
    <studioCategory>Data Controls</studioCategory>
    <helpUrl>https://docs.mendix.com/appstore/modules/data-grid-2#7-2-drop-down-filter</helpUrl>
    <properties>
        <propertyGroup caption="General">
            <propertyGroup caption="General">
                <property key="auto" type="boolean" defaultValue="true">
                    <caption>Automatic options</caption>
                    <description>Show options based on the references or the enumeration values and captions.</description>
                </property>
                <property key="advanced" type="boolean" defaultValue="false">
                    <caption>Enable advanced options</caption>
                    <description />
                </property>
                <property key="defaultValue" type="expression" required="false">
                    <caption>Default value</caption>
                    <description>Empty option caption will be shown by default or if configured default value matches none of the options</description>
                    <returnType type="String" />
                </property>
                <property key="filterOptions" type="object" isList="true" required="false">
                    <caption>Options</caption>
                    <description />
                    <properties>
                        <property key="caption" type="textTemplate">
                            <caption>Caption</caption>
                            <category>General</category>
                            <description />
                        </property>
                        <property key="value" type="expression">
                            <caption>Value</caption>
                            <category>General</category>
                            <description />
                            <returnType type="String" />
                        </property>
                    </properties>
                </property>
                <property key="emptyOptionCaption" type="textTemplate" required="false">
                    <caption>Empty option caption</caption>
                    <description />
                </property>
                <property key="multiSelect" type="boolean" defaultValue="false">
                    <caption>Multiselect</caption>
                    <description />
                </property>
            </propertyGroup>
            <propertyGroup caption="Configurations">
                <property key="valueAttribute" type="attribute" required="false">
                    <caption>Saved attribute</caption>
                    <description>Attribute used to store the last value of the filter. Associations are not supported.</description>
                    <attributeTypes>
                        <attributeType name="String" />
                    </attributeTypes>
                </property>
            </propertyGroup>
            <propertyGroup caption="Events">
                <property key="onChange" type="action" required="false">
                    <caption>On change</caption>
                    <description>Action to be triggered when the value or filter changes.</description>
                </property>
            </propertyGroup>
        </propertyGroup>
        <propertyGroup caption="Accessibility">
            <propertyGroup caption="Accessibility">
                <property key="ariaLabel" type="textTemplate" required="false">
                    <caption>Input caption</caption>
                    <description>Assistive technology will read this upon reaching the input element.</description>
                </property>
            </propertyGroup>
        </propertyGroup>
    </properties>
</widget>
