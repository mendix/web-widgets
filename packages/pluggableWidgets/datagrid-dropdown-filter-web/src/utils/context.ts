import { FilterContextValue } from "@mendix/pluggable-widgets-commons/dist/components/web";

export function getProviderChannel(context: FilterContextValue): string | undefined {
    return context.providerData?.eventChannelName;
}
