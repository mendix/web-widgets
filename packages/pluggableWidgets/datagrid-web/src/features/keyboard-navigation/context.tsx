import { createContext, useMemo, PropsWithChildren, ReactElement, createElement, useContext } from "react";
import { FocusTargetController } from "./FocusTargetController";
import { useFocusTargetController, LayoutProps } from "./useFocusTargetController";

export const defaultValue = Symbol("DefaultKeyNavContextValue");

export type ContextValue = {
    focusController: FocusTargetController;
};

export const context = createContext<typeof defaultValue | ContextValue>(defaultValue);

export function KeyNavProvider(props: PropsWithChildren<LayoutProps>): ReactElement | null {
    const focusController = useFocusTargetController(props);
    const value: ContextValue = useMemo(() => ({ focusController }), [focusController]);

    return <context.Provider value={value}>{props.children}</context.Provider>;
}

export function useKeyNavContext(): ContextValue {
    const contextValue = useContext(context);

    if (contextValue === defaultValue) {
        throw new Error("useKeyNavContext: context provider is missing");
    }

    return contextValue;
}
