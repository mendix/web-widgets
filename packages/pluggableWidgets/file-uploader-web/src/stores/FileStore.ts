import { Big } from "big.js";
import { ObjectItem } from "mendix";
import { action, computed, makeObservable, observable, runInAction } from "mobx";
import mimeTypes from "mime-types";

import { FileUploaderStore } from "./FileUploaderStore";
import {
    fetchDocumentUrl,
    fetchImageThumbnail,
    fetchMxObject,
    isImageObject,
    MxObject,
    removeObject,
    saveFile
} from "../utils/mx-data";

export type FileStatus =
    | "existingFile"
    | "new"
    | "uploading"
    | "done"
    | "uploadingError"
    | "removedFile"
    | "validationError";

let fileKey = 0;

function getFileKey(): number {
    return fileKey++;
}

export class FileStore {
    fileStatus: FileStatus;

    _file?: File = undefined;
    _objectItem?: ObjectItem = undefined;
    _mxObject?: MxObject = undefined;
    _thumbnailUrl?: string = undefined;
    _rootStore: FileUploaderStore;

    key: number;

    errorDescription?: string = undefined;

    constructor(type: FileStatus, rootStore: FileUploaderStore, file?: File, objectItem?: ObjectItem) {
        this.key = getFileKey();
        this._file = file;
        this._objectItem = objectItem;
        this._rootStore = rootStore;
        this.fileStatus = type;

        makeObservable(this, {
            fileStatus: observable,
            _mxObject: observable,
            errorDescription: observable,
            _thumbnailUrl: observable,
            canRemove: computed,
            imagePreviewUrl: computed,
            upload: action,
            fetchMxObject: action
        });
    }

    validate(): boolean {
        return !(this.fileStatus !== "new" || !this._file);
    }

    async upload(): Promise<void> {
        if (this.fileStatus === "existingFile") {
            throw new Error("Calling upload on already uploaded files is not supported");
        }

        // set status
        runInAction(() => {
            this.fileStatus = "uploading";
        });

        try {
            // request object item
            this._objectItem = await this._rootStore.objectCreationHelper.request();
            await saveFile(this._objectItem, this._file!);
            await this.fetchMxObject();

            runInAction(() => {
                this.fileStatus = "done";
            });
        } catch (e: unknown) {
            runInAction(() => {
                this.fileStatus = "uploadingError";
            });
        }
    }

    get title(): string {
        if (this._mxObject) {
            return this._mxObject?.get2("Name").toString();
        }

        return this._file?.name ?? "...";
    }

    get size(): number {
        if (this._mxObject) {
            return (this._mxObject.get2("Size") as Big).toNumber();
        }

        return this._file?.size ?? -1;
    }

    get mimeType(): string {
        return mimeTypes.lookup(this.title) || "application/octet-stream";
    }

    get canRemove(): boolean {
        return (!this._rootStore.isReadOnly && this.fileStatus === "existingFile") || this.fileStatus === "done";
    }

    get canDownload(): boolean {
        return this.fileStatus === "done" || this.fileStatus === "existingFile";
    }

    async remove(): Promise<void> {
        if (!this.canRemove || !this._objectItem) {
            return;
        }

        try {
            await removeObject(this._objectItem);
            runInAction(() => {
                this.fileStatus = "removedFile";
                this._mxObject = undefined;
                this.updateThumbnailUrl();
            });
        } catch (e: unknown) {
            console.error("Could not remove object:", e);
        }
    }

    async fetchMxObject(): Promise<void> {
        if (this._objectItem) {
            const obj = await fetchMxObject(this._objectItem!);
            runInAction(() => {
                this._mxObject = obj;
                this.updateThumbnailUrl();
            });
        }
    }

    async updateThumbnailUrl(): Promise<void> {
        if (this._rootStore._uploadMode !== "images") {
            return;
        }

        this._thumbnailUrl = undefined;
        if (this._mxObject && isImageObject(this._mxObject)) {
            const url = await fetchImageThumbnail(this._mxObject);
            runInAction(() => {
                this._thumbnailUrl = url;
            });
        }
    }

    get downloadUrl(): string | undefined {
        if (this._mxObject) {
            return fetchDocumentUrl(this._mxObject);
        }
    }

    get imagePreviewUrl(): string | undefined {
        if (this._rootStore._uploadMode !== "images") {
            return;
        }

        if (this._thumbnailUrl) {
            return this._thumbnailUrl;
        }
        if (this._file && this.mimeType.startsWith("image/")) {
            return URL.createObjectURL(this._file!);
        }

        return undefined;
    }

    static existingFile(objectItem: ObjectItem, rootStore: FileUploaderStore): FileStore {
        const store = new FileStore("existingFile", rootStore, undefined, objectItem);
        store.fetchMxObject();

        return store;
    }

    static newFile(file: File, rootStore: FileUploaderStore): FileStore {
        return new FileStore("new", rootStore, file, undefined);
    }

    static newFileWithError(file: File, errorMessage: string, rootStore: FileUploaderStore): FileStore {
        const store = new FileStore("validationError", rootStore, file, undefined);
        runInAction(() => {
            store.errorDescription = errorMessage;
        });

        return store;
    }
}
