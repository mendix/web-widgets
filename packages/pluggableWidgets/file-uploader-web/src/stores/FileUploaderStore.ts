import { ActionValue, ListValue, ObjectItem } from "mendix";
import { FileUploaderContainerProps, UploadModeEnum } from "../../typings/FileUploaderProps";
import { action, computed, makeObservable, observable } from "mobx";
import { getImageUploaderFormats, parseAllowedFormats } from "../utils/parseAllowedFormats";
import { FileStore } from "./FileStore";
import { fileHasContents } from "../utils/mx-data";
import { FileRejection } from "react-dropzone";
import { FileCheckFormat } from "../utils/predefinedFormats";
import { TranslationsStore } from "./TranslationsStore";

export class FileUploaderStore {
    files: FileStore[] = [];
    lastSeenItems: Set<ObjectItem["id"]> = new Set<ObjectItem["id"]>();
    currentWaiting: Array<(v: ObjectItem | undefined) => void> = [];
    itemCreationTimeout?: number = undefined;

    existingItemsLoaded = false;
    isReadOnly: boolean;

    acceptedFileTypes: FileCheckFormat[];

    _widgetName: string;
    _uploadMode: UploadModeEnum;
    _createObjectAction?: ActionValue;
    _maxFileSizeMb = 0;
    _maxFileSize = 0;
    _ds?: ListValue;
    _maxFilesPerUpload: number;
    _objectCreationTimeout: number;

    errorMessage?: string = undefined;

    translations: TranslationsStore;

    constructor(props: FileUploaderContainerProps, translations: TranslationsStore) {
        this._widgetName = props.name;
        this._maxFileSizeMb = props.maxFileSize;
        this._maxFileSize = this._maxFileSizeMb * 1024 * 1024;
        this._maxFilesPerUpload = props.maxFilesPerUpload;
        this._uploadMode = props.uploadMode;
        this._objectCreationTimeout = props.objectCreationTimeout;

        this.isReadOnly = props.readOnlyMode;

        this.acceptedFileTypes =
            this._uploadMode === "files" ? parseAllowedFormats(props.allowedFileFormats) : getImageUploaderFormats();

        this.translations = translations;

        makeObservable(this, {
            updateProps: action,
            processDrop: action,
            setMessage: action,
            processExistingFileItem: action,
            files: observable,
            existingItemsLoaded: observable,
            errorMessage: observable,
            allowedFormatsDescription: computed
        });

        this.updateProps(props);
    }

    updateProps(props: FileUploaderContainerProps): void {
        if (props.uploadMode === "files") {
            this._createObjectAction = props.createFileAction;
            this._ds = props.associatedFiles;
        } else {
            this._createObjectAction = props.createImageAction;
            this._ds = props.associatedImages;
        }

        this.translations.updateProps(props);

        const itemsDs = this._ds;
        if (!this.existingItemsLoaded) {
            if (itemsDs.status === "available" && itemsDs.items) {
                for (const item of itemsDs.items) {
                    this.processExistingFileItem(item);
                }

                this.existingItemsLoaded = true;
            }
        } else {
            for (const newItem of findNewItems(this.lastSeenItems, itemsDs.items || [])) {
                if (!fileHasContents(newItem)) {
                    this.processEmptyFileItem(newItem);
                } else {
                    // adding this file to the list as is as this file is not empty and probably created externally
                    this.processExistingFileItem(newItem);
                }
            }
        }
    }

    processExistingFileItem(item: ObjectItem): void {
        this.files.unshift(FileStore.existingFile(item, this));

        this.lastSeenItems.add(item.id);
    }

    processEmptyFileItem(item: ObjectItem): void {
        if (this.isReadOnly) {
            return;
        }
        const firstWaiting = this.currentWaiting.shift();
        if (firstWaiting) {
            firstWaiting(item);
        }

        this.lastSeenItems.add(item.id);

        this.executeFileObjectCreation();
    }

    get allowedFormatsDescription(): string {
        return this.acceptedFileTypes
            .map(f => {
                return f.description;
            })
            .filter(_ => _)
            .join(", ");
    }

    get canRequestFile(): boolean {
        return this.existingItemsLoaded && !!this._createObjectAction;
    }

    requestFileObject(): Promise<ObjectItem | undefined> {
        if (!this.canRequestFile) {
            throw new Error("Can't request file");
        }

        return new Promise<ObjectItem | undefined>(resolve => {
            this.currentWaiting.push(resolve);

            this.executeFileObjectCreation();
        });
    }

    executeFileObjectCreation(): void {
        clearTimeout(this.itemCreationTimeout);
        this.itemCreationTimeout = undefined;

        if (!this.currentWaiting.length) {
            return;
        }
        // we need to check if creation is taking too much time
        // start the timer to measure how long it takes,
        // if a threshold is reached, declare it a failure
        // this means the action is probably misconfigured.
        this.itemCreationTimeout = setTimeout(() => {
            console.error(
                `Looks like the 'Action to create new files/images' action did not create any objects within ${this._objectCreationTimeout} seconds. Please check if '${this._widgetName}' widget is configured correctly.`
            );
            // fail all waiting
            while (this.currentWaiting.length) {
                this.currentWaiting.shift()?.(undefined);
            }
        }, this._objectCreationTimeout * 1000) as any as number;

        this._createObjectAction!.execute();
    }

    setMessage(msg?: string): void {
        this.errorMessage = msg;
    }

    processDrop(acceptedFiles: File[], fileRejections: FileRejection[]): void {
        if (!this._createObjectAction || !this._createObjectAction.canExecute) {
            console.error(
                `'Action to create new files/images' is not available or can't be executed. Please check if '${this._widgetName}' widget is configured correctly.`
            );
            this.setMessage(this.translations.get("unavailableCreateActionMessage"));
            return;
        }

        if (fileRejections.length && fileRejections[0].errors[0].code === "too-many-files") {
            this.setMessage(
                this.translations.get("uploadFailureTooManyFilesMessage", this._maxFilesPerUpload.toString())
            );
            return;
        }

        this.setMessage();

        for (const file of fileRejections) {
            const newFileStore = FileStore.newFileWithError(
                file.file,
                file.errors
                    .map(e => {
                        if (e.code === "file-invalid-type") {
                            return this.translations.get(
                                "uploadFailureInvalidFileFormatMessage",
                                this.allowedFormatsDescription
                            );
                        }
                        if (e.code === "file-too-large") {
                            return this.translations.get(
                                "uploadFailureFileIsTooBigMessage",
                                this._maxFileSizeMb.toString()
                            );
                        }
                        return e.message;
                    })
                    .join(" "),
                this
            );

            this.files.unshift(newFileStore);
        }

        for (const file of acceptedFiles) {
            const newFileStore = FileStore.newFile(file, this);

            this.files.unshift(newFileStore);

            if (newFileStore.validate()) {
                newFileStore.upload();
            }
        }
    }
}

function findNewItems(lastSeenItems: Set<ObjectItem["id"]>, currentItems: ObjectItem[]): ObjectItem[] {
    return currentItems.filter(i => !lastSeenItems.has(i.id));
}
