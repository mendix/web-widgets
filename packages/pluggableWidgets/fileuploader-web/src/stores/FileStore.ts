import { Big } from "big.js";
import { ObjectItem } from "mendix";
import { action, computed, makeObservable, observable, runInAction } from "mobx";

import { FileUploaderStore } from "./FileUploaderStore";
import { fetchMxObject, MxObject, removeObject, saveFile } from "../utils/mx-data";

export type FileStatus =
    | "existingFile"
    | "new"
    | "uploading"
    | "done"
    | "uploadingError"
    | "removedFile"
    | "validationError";

let fileKey = 0;

function getFileKey(): number {
    return fileKey++;
}

export class FileStore {
    fileStatus: FileStatus;

    file?: File = undefined;
    objectItem?: ObjectItem = undefined;
    mxObject?: MxObject = undefined;
    rootStore: FileUploaderStore;

    key: number;

    errorDescription?: string = undefined;

    constructor(type: FileStatus, rootStore: FileUploaderStore, file?: File, objectItem?: ObjectItem) {
        this.key = getFileKey();
        this.file = file;
        this.objectItem = objectItem;
        this.rootStore = rootStore;
        this.fileStatus = type;

        makeObservable(this, {
            fileStatus: observable,
            mxObject: observable,
            errorDescription: observable,
            canRemove: computed,
            upload: action,
            fetchMxObject: action
        });
    }

    validate(): boolean {
        return !(this.fileStatus !== "new" || !this.file);
    }

    async upload(): Promise<void> {
        if (this.fileStatus === "existingFile") {
            throw new Error("Calling upload on already uploaded files is not supported");
        }

        // set status
        runInAction(() => {
            this.fileStatus = "uploading";
        });

        try {
            // request object item
            this.objectItem = await this.rootStore.requestFileObject();
            if (this.objectItem) {
                await saveFile(this.objectItem, this.file!);
                await this.fetchMxObject();

                runInAction(() => {
                    this.fileStatus = "done";
                });
            } else {
                runInAction(() => {
                    this.fileStatus = "uploadingError";
                });
            }
        } catch (e: unknown) {
            runInAction(() => {
                this.fileStatus = "uploadingError";
            });
        }
    }

    get title(): string {
        if (this.mxObject) {
            return this.mxObject?.get2("Name").toString();
        }

        return this.file?.name ?? "...";
    }

    get size(): number {
        if (this.mxObject) {
            return (this.mxObject.get2("Size") as Big).toNumber();
        }

        return this.file?.size ?? -1;
    }

    get canRemove(): boolean {
        return this.fileStatus === "existingFile" || this.fileStatus === "done";
    }

    async remove(): Promise<void> {
        if (!this.canRemove || !this.objectItem) {
            return;
        }

        try {
            await removeObject(this.objectItem);
            runInAction(() => {
                this.fileStatus = "removedFile";
                this.mxObject = undefined;
            });
        } catch (e: unknown) {
            console.error("Could not remove object:", e);
        }
    }

    async fetchMxObject(): Promise<void> {
        if (this.objectItem) {
            const obj = await fetchMxObject(this.objectItem!);
            runInAction(() => {
                this.mxObject = obj;
            });
        }
    }

    static existingFile(objectItem: ObjectItem, rootStore: FileUploaderStore): FileStore {
        const store = new FileStore("existingFile", rootStore, undefined, objectItem);
        store.fetchMxObject();

        return store;
    }

    static newFile(file: File, rootStore: FileUploaderStore): FileStore {
        return new FileStore("new", rootStore, file, undefined);
    }

    static newFileWithError(file: File, errorMessage: string, rootStore: FileUploaderStore): FileStore {
        const store = new FileStore("validationError", rootStore, file, undefined);
        runInAction(() => {
            store.errorDescription = errorMessage;
        });

        return store;
    }
}
