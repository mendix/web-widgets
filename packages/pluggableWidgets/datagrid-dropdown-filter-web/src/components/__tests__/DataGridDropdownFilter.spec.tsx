import "@testing-library/jest-dom";
import { FilterAPIv2 } from "@mendix/widget-plugin-filtering/context";
import { HeaderFiltersStore, HeaderFiltersStoreProps } from "@mendix/widget-plugin-filtering/stores/HeaderFiltersStore";
import { dynamicValue, ListAttributeValueBuilder } from "@mendix/widget-plugin-test-utils";
import { createContext, createElement } from "react";
import DatagridDropdownFilter from "../../DatagridDropdownFilter";
import { fireEvent, render, screen, waitFor } from "@testing-library/react";

const commonProps = {
    class: "filter-custom-class",
    tabIndex: 0,
    name: "filter-test",
    advanced: false,
    groupKey: "dropdown-filter"
};

describe("Dropdown Filter", () => {
    describe("with single instance", () => {
        afterEach(() => {
            delete (global as any)["com.mendix.widgets.web.UUID"];
        });

        describe("with single attribute", () => {
            function mockCtx(universe: string[]): void {
                const props: HeaderFiltersStoreProps = {
                    filterList: [
                        {
                            filter: new ListAttributeValueBuilder()
                                .withUniverse(universe)
                                .withType("Enum")
                                .withFilterable(true)
                                .withFormatter(
                                    value => value,
                                    () => console.log("Parsed")
                                )
                                .build()
                        }
                    ]
                };
                const headerFilterStore = new HeaderFiltersStore(props, null);
                (window as any)["com.mendix.widgets.web.filterable.filterContext.v2"] = createContext<FilterAPIv2>(
                    headerFilterStore.context
                );
            }
            beforeEach(() => {
                mockCtx(["enum_value_1", "enum_value_2"]);
            });

            describe("with auto options", () => {
                it("loads correct values from universe", async () => {
                    const filter = render(
                        <DatagridDropdownFilter {...commonProps} auto multiSelect={false} filterOptions={[]} />
                    );

                    const trigger = filter.getByRole("textbox");

                    await fireEvent.click(trigger);

                    const items = filter.getAllByRole("menuitem");

                    items.forEach((item, index) => {
                        if (index === 0) {
                            return;
                        }
                        expect(item.textContent).toEqual(`enum_value_${index}`);
                    });
                });
            });

            describe("DOM structure", () => {
                it("renders correctly", () => {
                    const { asFragment } = render(
                        <DatagridDropdownFilter {...commonProps} auto multiSelect={false} filterOptions={[]} />
                    );

                    expect(asFragment()).toMatchSnapshot();
                });
            });

            describe("with defaultValue", () => {
                it("initialize component with defaultValue", () => {
                    render(
                        <DatagridDropdownFilter
                            {...commonProps}
                            auto
                            multiSelect={false}
                            filterOptions={[]}
                            defaultValue={dynamicValue<string>("enum_value_1")}
                        />
                    );

                    expect(screen.getByRole("textbox")).toHaveValue("enum_value_1");
                });

                it("don't sync defaultValue with state when defaultValue changes from undefined to string", async () => {
                    const { rerender } = render(
                        <DatagridDropdownFilter
                            {...commonProps}
                            auto
                            multiSelect={false}
                            filterOptions={[]}
                            defaultValue={dynamicValue<string>("")}
                        />
                    );

                    await waitFor(() => {
                        expect(screen.getByRole("textbox")).toHaveValue("");
                    });

                    // “Real” context causes widgets to re-renders multiple times, replicate this in mocked context.
                    rerender(
                        <DatagridDropdownFilter
                            {...commonProps}
                            auto
                            multiSelect={false}
                            filterOptions={[]}
                            defaultValue={dynamicValue<string>("")}
                        />
                    );
                    rerender(
                        <DatagridDropdownFilter
                            {...commonProps}
                            auto
                            multiSelect={false}
                            filterOptions={[]}
                            defaultValue={dynamicValue<string>("enum_value_1")}
                        />
                    );

                    await waitFor(() => {
                        expect(screen.getByRole("textbox")).toHaveValue("");
                    });
                });

                it("don't sync defaultValue with state when defaultValue changes from string to undefined", async () => {
                    mockCtx(["xyz", "abc"]);
                    const { rerender } = render(
                        <DatagridDropdownFilter
                            {...commonProps}
                            auto
                            multiSelect={false}
                            filterOptions={[]}
                            defaultValue={dynamicValue<string>("xyz")}
                        />
                    );

                    expect(screen.getByRole("textbox")).toHaveValue("xyz");

                    // “Real” context causes widgets to re-renders multiple times, replicate this in mocked context.
                    rerender(
                        <DatagridDropdownFilter
                            {...commonProps}
                            auto
                            multiSelect={false}
                            filterOptions={[]}
                            defaultValue={dynamicValue<string>("xyz")}
                        />
                    );
                    rerender(
                        <DatagridDropdownFilter
                            {...commonProps}
                            auto
                            multiSelect={false}
                            filterOptions={[]}
                            defaultValue={undefined}
                        />
                    );

                    await waitFor(() => {
                        expect(screen.getByRole("textbox")).toHaveValue("xyz");
                    });
                });
            });

            afterAll(() => {
                (window as any)["com.mendix.widgets.web.filterable.filterContext.v2"] = undefined;
            });
        });

        describe("with multiple attributes", () => {
            beforeAll(() => {
                const props: HeaderFiltersStoreProps = {
                    filterList: [
                        {
                            filter: new ListAttributeValueBuilder()
                                .withId("attribute1")
                                .withUniverse(["enum_value_1", "enum_value_2"])
                                .withType("Enum")
                                .withFilterable(true)
                                .withFormatter(
                                    value => value,
                                    () => console.log("Parsed")
                                )
                                .build()
                        },
                        {
                            filter: new ListAttributeValueBuilder()
                                .withId("attribute2")
                                .withUniverse([true, false])
                                .withType("Boolean")
                                .withFilterable(true)
                                .withFormatter(
                                    value => (value ? "Yes" : "No"),
                                    () => console.log("Parsed")
                                )
                                .build()
                        }
                    ]
                };
                const headerFilterStore = new HeaderFiltersStore(props, null);
                (window as any)["com.mendix.widgets.web.filterable.filterContext.v2"] = createContext<FilterAPIv2>(
                    headerFilterStore.context
                );
            });

            describe("with auto options", () => {
                it("loads correct values from universes", async () => {
                    const filter = render(
                        <DatagridDropdownFilter {...commonProps} auto multiSelect={false} filterOptions={[]} />
                    );

                    const trigger = filter.getByRole("textbox");
                    await fireEvent.click(trigger);

                    expect(filter.getAllByRole("menuitem").map(item => item.textContent)).toStrictEqual([
                        "",
                        "enum_value_1",
                        "enum_value_2",
                        "Yes",
                        "No"
                    ]);
                });
            });

            afterAll(() => {
                (window as any)["com.mendix.widgets.web.filterable.filterContext.v2"] = undefined;
            });
        });

        describe("with wrong attribute's type", () => {
            beforeAll(() => {
                const props: HeaderFiltersStoreProps = {
                    filterList: [
                        {
                            filter: new ListAttributeValueBuilder().withType("String").withFilterable(true).build()
                        }
                    ]
                };
                const headerFilterStore = new HeaderFiltersStore(props, null);
                (window as any)["com.mendix.widgets.web.filterable.filterContext.v2"] = createContext<FilterAPIv2>(
                    headerFilterStore.context
                );
            });

            it("renders error message", () => {
                const { container } = render(
                    <DatagridDropdownFilter {...commonProps} auto multiSelect={false} filterOptions={[]} />
                );

                expect(container.querySelector(".alert")?.textContent).toBe(
                    "Unable to get filter store. Check parent widget configuration."
                );
            });

            afterAll(() => {
                (window as any)["com.mendix.widgets.web.filterable.filterContext.v2"] = undefined;
            });
        });

        describe("with wrong multiple attributes' types", () => {
            beforeAll(() => {
                const props: HeaderFiltersStoreProps = {
                    filterList: [
                        {
                            filter: new ListAttributeValueBuilder()
                                .withId("attribute1")
                                .withType("String")
                                .withFilterable(true)
                                .build()
                        },
                        {
                            filter: new ListAttributeValueBuilder()
                                .withId("attribute2")
                                .withType("Decimal")
                                .withFilterable(true)
                                .build()
                        }
                    ]
                };
                const headerFilterStore = new HeaderFiltersStore(props, null);
                (window as any)["com.mendix.widgets.web.filterable.filterContext.v2"] = createContext<FilterAPIv2>(
                    headerFilterStore.context
                );
            });

            it("renders error message", () => {
                const { container } = render(
                    <DatagridDropdownFilter {...commonProps} auto multiSelect={false} filterOptions={[]} />
                );

                expect(container.querySelector(".alert")?.textContent).toBe(
                    "Unable to get filter store. Check parent widget configuration."
                );
            });

            afterAll(() => {
                (window as any)["com.mendix.widgets.web.filterable.filterContext.v2"] = undefined;
            });
        });

        describe("with no context", () => {
            beforeAll(() => {
                (window as any)["com.mendix.widgets.web.filterable.filterContext.v2"] = undefined;
            });

            it("renders error message", () => {
                const { container } = render(
                    <DatagridDropdownFilter {...commonProps} auto multiSelect={false} filterOptions={[]} />
                );

                expect(container.querySelector(".alert")?.textContent).toBe(
                    "The filter widget must be placed inside the column or header of the Data grid 2.0 or inside header of the Gallery widget."
                );
            });
        });

        describe("with invalid values", () => {
            beforeAll(() => {
                const props: HeaderFiltersStoreProps = {
                    filterList: [
                        {
                            filter: new ListAttributeValueBuilder()
                                .withUniverse(["enum_value_1", "enum_value_2"])
                                .withType("Enum")
                                .withFilterable(true)
                                .build()
                        }
                    ]
                };
                const headerFilterStore = new HeaderFiltersStore(props, null);
                (window as any)["com.mendix.widgets.web.filterable.filterContext.v2"] = createContext<FilterAPIv2>(
                    headerFilterStore.context
                );
            });

            it("renders error message", () => {
                const { container } = render(
                    <DatagridDropdownFilter
                        {...commonProps}
                        auto={false}
                        multiSelect={false}
                        filterOptions={[
                            {
                                caption: dynamicValue<string>("wrong value"),
                                value: dynamicValue<string>("enum_value_3")
                            }
                        ]}
                    />
                );

                expect(container.querySelector(".alert")?.textContent).toBe("Invalid option value: 'enum_value_3'");
            });
        });

        describe("with multiple invalid values", () => {
            beforeAll(() => {
                const props: HeaderFiltersStoreProps = {
                    filterList: [
                        {
                            filter: new ListAttributeValueBuilder()
                                .withUniverse(["enum_value_1", "enum_value_2"])
                                .withType("Enum")
                                .withFilterable(true)
                                .build()
                        },
                        {
                            filter: new ListAttributeValueBuilder()
                                .withUniverse([true, false])
                                .withType("Boolean")
                                .withFilterable(true)
                                .build()
                        }
                    ]
                };
                const headerFilterStore = new HeaderFiltersStore(props, null);
                (window as any)["com.mendix.widgets.web.filterable.filterContext.v2"] = createContext<FilterAPIv2>(
                    headerFilterStore.context
                );
            });

            it("renders error message", () => {
                const { container } = render(
                    <DatagridDropdownFilter
                        {...commonProps}
                        auto={false}
                        multiSelect={false}
                        filterOptions={[
                            {
                                caption: dynamicValue<string>("wrong enum value"),
                                value: dynamicValue<string>("enum_value_3")
                            },
                            {
                                caption: dynamicValue<string>("wrong boolean value"),
                                value: dynamicValue<string>("no")
                            }
                        ]}
                    />
                );

                expect(container.querySelector(".alert")?.textContent).toBe("Invalid option value: 'enum_value_3'");
            });
        });
    });

    describe("with multiple instances", () => {
        beforeAll(() => {
            const props: HeaderFiltersStoreProps = {
                filterList: [
                    {
                        filter: new ListAttributeValueBuilder()
                            .withUniverse(["enum_value_1", "enum_value_2"])
                            .withType("Enum")
                            .withFilterable(true)
                            .withFormatter(
                                value => value,
                                () => console.log("Parsed")
                            )
                            .build()
                    }
                ]
            };
            const headerFilterStore = new HeaderFiltersStore(props, null);
            (window as any)["com.mendix.widgets.web.filterable.filterContext.v2"] = createContext<FilterAPIv2>(
                headerFilterStore.context
            );
        });

        it("renders with a unique id", () => {
            const { asFragment: fragment1 } = render(
                <DatagridDropdownFilter {...commonProps} auto multiSelect={false} filterOptions={[]} />
            );
            const { asFragment: fragment2 } = render(
                <DatagridDropdownFilter {...commonProps} auto multiSelect={false} filterOptions={[]} />
            );

            expect(fragment1().querySelector("input")?.getAttribute("aria-controls")).not.toBe(
                fragment2().querySelector("input")?.getAttribute("aria-controls")
            );
        });

        afterAll(() => {
            (window as any)["com.mendix.widgets.web.filterable.filterContext.v2"] = undefined;
            delete (global as any)["com.mendix.widgets.web.UUID"];
        });
    });
});
