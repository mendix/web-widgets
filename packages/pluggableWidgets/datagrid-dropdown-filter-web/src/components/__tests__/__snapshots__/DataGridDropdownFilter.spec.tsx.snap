// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Dropdown Filter with single instance with single attribute DOM structure renders correctly 1`] = `
<DocumentFragment>
  <div
    class="dropdown-container filter-custom-class"
    data-focusindex="0"
  >
    <input
      aria-controls="Dropdown1-dropdown-list"
      aria-expanded="false"
      aria-haspopup="true"
      class="form-control dropdown-triggerer"
      placeholder=""
      value=""
    />
  </div>
</DocumentFragment>
`;
